feat: allow for async parsing in file loader (#14932)
withastro/astro: .changeset/cozy-masks-count.md

---
'astro': minor
---

> Adds support for returning a Promise from the `parser()` option of the `file()` loader. This enables you to run asynchronous code such as fetching remote data or using async parsers, when loading files with the Content Layer API.

Reviewer: ```suggestion
Adds support for returning a Promise from the `parser()` option of the `file()` loader

This enables you to run asynchronous code such as fetching remote data or using async parsers when loading files with the Content Layer API.
```
(Longer changelog messages are easier to read when we stick to one line for the main change, because they're part of the whole PR number/thanks author etc. and that gets big quickly, so I'd typically do this)

For example:

```js
import { defineCollection } from 'astro:content';
import { file } from 'astro/loaders';

const blog = defineCollection({
  loader: file('src/data/blog.json', {
    parser: async (text) => {
      const data = JSON.parse(text);
      
      // Perform async operations like fetching additional data
      const enrichedData = await fetch(`https://api.example.com/enrich`, {
        method: 'POST',
        body: JSON.stringify(data),
      }).then(res => res.json());
      
      return enrichedData;
    },
  }),
});

export const collections = { blog };

> ```

Reviewer: ``````suggestion
```

See [the `parser()` reference documentation](https://docs.astro.build/en/reference/content-loader-reference/#parser) for more information.
``````
If we have a good docs link, I do like to throw it in at the end!

---
'astro': minor
---

> Allow the `parser()` option on the `file()` loader to return a Promise

Reviewer: For a minor release of a new feature, we'd usually hype it up a bit more, focusing on what a user can now do (with an example usage shown) rather than implementation details.  You can [see a model of this in our contributing docs](https://contribute.docs.astro.build/docs-for-code-changes/changesets/#new-features):

- Quick sentence to shout "New! Shiny!" and grab people's attention without a lot of visual clutter
- describe the problem it solves or ability it unlocks and why that's cool
- show an example that people could copy into their projects right away to try it out so that it's easy for them to get started
- link to the docs for the new thing

The first sentence could be something like...

```suggestion
Adds the ability to parse files in the `file()` loader asynchronously
```

```suggestion
Adds the option to define your `file()` loader  with an async `parser()` function to support additional data processing
```

Something along those lines, whatever makes sense?

Then, we'd want to explain a tiny bit why this is helpful, give an example of when its used and show this in use with code. (Looking at the tests, it seems you'd need to configure `parser()` as async, then await your helper data function in your file?) The model listed in the contributing docs should be a good example of a model you can follow!
